pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,858 @@
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.NewTypedItemExponentialFailureRateLimiter[string](minRetryDelay, maxRetryDelay),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "packetcapture"},
+		),
+		sftpUploader:      &ftp.SftpUploader{},
+		cond:              sync.NewCond(&sync.Mutex{}),
+		captures:          make(map[string]packetCapturePhase),
+		waitingCapturesCh: make(chan string, maxWaitingCaptures),
+	}
+
+	packetCaptureInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addPacketCapture,
+		UpdateFunc: c.updatePacketCapture,
+		DeleteFunc: c.deletePacketCapture,
+	}, resyncPeriod)
+
+	c.newPacketSourceFn = func(c *Controller, pc *crdv1alpha1.PacketCapture) (PacketSource, error) {
+		matchPacket, err := c.createMatchPacket(pc)
+		if err != nil {
+			return nil, err
+		}
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil, nil
+		}
+		filter := bpf.CompilePacketFilter(pc.Spec.Packet, matchPacket)
+		klog.V(5).InfoS("Generated bpf instructions for Packetcapture", "name", pc.Name, "match packet", matchPacket, "bpf instructions", filter)
+		return NewPcapSource(*device, filter)
+	}
+
+	return c
+}
+
+func (c *Controller) enqueuePacketCapture(pc *crdv1alpha1.PacketCapture) {
+	c.queue.Add(pc.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the PacketCapture events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting packetcapture controller", "name", controllerName)
+	defer klog.InfoS("Shutting down packetcapture controller", "name", controllerName)
+
+	cacheSynced := []cache.InformerSynced{c.packetCaptureSynced}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSynced...) {
+		return
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
+		klog.ErrorS(err, "Couldn't create the directory for storing captured packets", "directory", packetDirectory)
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	// check the captures that are waiting in line.
+	go c.processWaitingCaptures()
```

Comment:
I don't think this goroutine should be created more than once?

Reply:
moved to `Run`.